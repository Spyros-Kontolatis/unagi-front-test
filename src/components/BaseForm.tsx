import React, { useState, useEffect } from 'react';
import dayjs, { Dayjs } from 'dayjs';

import { FormControl, TextField, Box, Button } from '@mui/material';

import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { DatePicker } from '@mui/x-date-pickers';

import type { Field, FieldValues } from '../types';

export default ({
  fields,
  onCancel,
  onSubmit,
}: {
  fields: Field[];
  onCancel: () => void;
  onSubmit: (values: FieldValues) => void;
}) => {
  const [values, setValues] = useState<FieldValues>(
    fields.reduce((acc, curr) => {
      if (curr.type === 'datepicker') {
        acc[curr.key] = curr.initialValue
          ? curr.initialValue
          : dayjs(new Date());
      } else {
        acc[curr.key] = curr.initialValue ? curr.initialValue : '';
      }
      return acc;
    }, {}),
  );

  useEffect(() => {
    setValues(
      fields.reduce((acc, curr) => {
        if (curr.type === 'datepicker') {
          acc[curr.key] = curr.initialValue
            ? dayjs(curr.initialValue)
            : dayjs(new Date());
        } else {
          acc[curr.key] = curr.initialValue ? curr.initialValue : '';
        }
        return acc;
      }, {}),
    );
  }, [fields]);

  const [errors, setErrors] = useState<FieldValues>(
    fields.reduce((acc, curr) => {
      acc[curr.key] = '';
      return acc;
    }, {}),
  );

  const validate = () => {
    let errorObj = {};
    fields.forEach((field) => {
      errorObj[field.key] = field.error?.(values[field.key]) ?? '';
    });
    setErrors(errorObj);
    return !Object.values(errorObj).some((err) => !!err);
  };

  const submit = () => {
    if (!validate()) return;
    onSubmit(values);
  };

  return (
    <Box
      component="div"
      sx={{
        display: 'flex',
        flexDirection: 'column',
        flexWrap: 'wrap',
        marginBottom: '8px',
        marginTop: '24px',
        justifyContent: 'center',
        alignItems: 'center',
        gap: '16px',
        width: '100%',
      }}
    >
      {fields?.map(({ type, label, key }: Field) => {
        if (type === 'textfield') {
          return (
            <FormControl key={key} sx={{ width: '70%' }}>
              <TextField
                id={key}
                label={label}
                variant="outlined"
                error={!!errors?.[key]}
                value={values?.[key]}
                onChange={(e: React.ChangeEvent) => {
                  const newValue = (e.target as HTMLInputElement).value;
                  setValues({ ...values, [key]: newValue });
                }}
                helperText={errors?.[key]}
              />
            </FormControl>
          );
        }

        if (type === 'datepicker') {
          return (
            <FormControl key={key} sx={{ width: '70%' }}>
              <LocalizationProvider dateAdapter={AdapterDayjs}>
                <DatePicker
                  label="Date of birth"
                  value={values?.[key] as Dayjs}
                  onChange={(newValue) => {
                    setValues({ ...values, [key]: newValue });
                  }}
                  slotProps={{
                    textField: {
                      error: !!errors?.[key],
                      helperText: errors?.[key],
                    },
                  }}
                />
              </LocalizationProvider>
            </FormControl>
          );
        }
      })}

      <Box
        sx={{
          width: '70%',
          display: 'flex',
          gap: '8px',
          flexWrap: 'wrap',
          justifyContent: 'center',
        }}
      >
        <Button color="error" variant="contained" onClick={onCancel}>
          Cancel
        </Button>
        <Button color="success" variant="contained" onClick={submit}>
          Submit
        </Button>
      </Box>
    </Box>
  );
};
