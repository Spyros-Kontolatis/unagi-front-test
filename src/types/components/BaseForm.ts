import type { Dayjs } from 'dayjs';
export interface Field {
  type: string;
  label: string;
  key: string;
  initialValue?: string | number | Dayjs;
  error: (val: string | number | Dayjs) => string;
}

export interface FieldValues {
  [key: string]: string | number | Dayjs;
}
