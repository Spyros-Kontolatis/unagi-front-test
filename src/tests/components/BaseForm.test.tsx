import BaseForm from '../../components/BaseForm';
import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';

describe('BaseCard', () => {
  test('renders the component', () => {
    const onCancel = jest.fn();
    const onSubmit = jest.fn();
    const { baseElement } = render(
      <BaseForm
        fields={[
          {
            type: 'textfield',
            label: 'First name',
            key: 'firstname',
            error: jest.fn(),
          },
          {
            type: 'textfield',
            label: 'Last name',
            key: 'lastname',
            error: jest.fn(),
          },
          {
            type: 'textfield',
            label: 'Image Url',
            key: 'image',
            error: jest.fn(),
          },
          {
            type: 'datepicker',
            label: 'Date of Birth',
            key: 'birthday',
            error: jest.fn(),
          },
        ]}
        onCancel={onCancel}
        onSubmit={onSubmit}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
  test('calls onCancel and onSubmit', () => {
    const onCancel = jest.fn();
    const onSubmit = jest.fn();
    const { baseElement } = render(
      <BaseForm
        fields={[
          {
            type: 'textfield',
            label: 'First name',
            key: 'firstname',
            error: jest.fn(),
          },
          {
            type: 'textfield',
            label: 'Last name',
            key: 'lastname',
            error: jest.fn(),
          },
          {
            type: 'textfield',
            label: 'Image Url',
            key: 'image',
            error: jest.fn(),
          },
          {
            type: 'datepicker',
            label: 'Date of Birth',
            key: 'birthday',
            error: jest.fn(),
          },
        ]}
        onCancel={onCancel}
        onSubmit={onSubmit}
      />,
    );
    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeTruthy();
    fireEvent.click(cancelButton);
    expect(onCancel).toHaveBeenCalledTimes(1);

    const submitButton = screen.getByText('Submit');
    expect(submitButton).toBeTruthy();
    fireEvent.click(submitButton);
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
