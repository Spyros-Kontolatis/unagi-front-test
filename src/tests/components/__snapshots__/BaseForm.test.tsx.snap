// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCard renders the component 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-iuoqln"
    >
      <div
        class="MuiFormControl-root css-12u03lg-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="firstname"
            id="firstname-label"
          >
            First name
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="firstname"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  First name
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
      <div
        class="MuiFormControl-root css-12u03lg-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="lastname"
            id="lastname-label"
          >
            Last name
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="lastname"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Last name
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
      <div
        class="MuiFormControl-root css-12u03lg-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="image"
            id="image-label"
          >
            Image Url
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="image"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Image Url
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
      <div
        class="MuiFormControl-root css-12u03lg-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-z3c6am-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-2"
            id="mui-2-label"
          >
            Date of birth
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-2"
              inputmode="text"
              placeholder="MM/DD/YYYY"
              type="text"
              value="06/15/2024"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <button
                aria-label="Choose date, selected date is Jun 15, 2024"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CalendarIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Date of birth
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
      <div
        class="MuiBox-root css-1ecqe9o"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorError MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorError css-1sszw6i-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess css-ke5b6m-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
